parser/impl/parser.cpp

```diff
@@ -469,17 +469,201 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
+                    m_errors.push_back(Error{
+                        .pos = m_lexer.Peek().pos,
+                        .message = "Expected to find ')'",
+                    });
+                    return nullptr;
+                }
+                t = m_lexer.Next(); // read ')'
+                expr.operand1 = std::make_shared<ast::RoutineCall>(rc);
+                return std::make_shared<ast::Expression>(expr);
+
+            } else if (t.type == TokenType::Dot ||
+                       t.type == TokenType::OpenBrack) {
+                // we can have a.b.c[7+9].d[0].a
+                // Identifier { . Identifier | [ Expression ] }
+                ast::ModifiablePrimary mp;
+                // ast::Primitive root;
+                // root.value = tok;
+                // expr.operation = t;
+
+                // mp.args.push_back(std::make_shared<ast::Expression>(root));
+
+                while (m_lexer.Peek().type == TokenType::Dot ||
+                       m_lexer.Peek().type == TokenType::OpenBrack) {
+
+                    t = m_lexer.Next(); // read either dot or '['
+                    ast::Primitive e;
+                    e.value = t;
+                    mp.args.push_back(std::make_shared<ast::Primitive>(e));
+                    if (t.type == TokenType::Dot) {
+                        // read, check and save identifier
+
+                        if (m_lexer.Peek().type != TokenType::Identifier) {
+                            m_errors.push_back(Error{
+                                .pos = m_lexer.Peek().pos,
+                                .message = "Expected to find an identifier",
+                            });
+                            return nullptr;
+                        }
+                        e.value = m_lexer.Next();
+                        mp.args.push_back(std::make_shared<ast::Primitive>(e));
+
+                    } else { // if []
+
+                        mp.args.push_back(parseBinaryExpression(0));
+
+                        // read and save expression
+                        if (m_lexer.Peek().type != TokenType::CloseBrack) {
+                            m_errors.push_back(Error{
+                                .pos = m_lexer.Peek().pos,
+                                .message = "Expected to find a ']' token",
+                            });
+                            return nullptr;
+                        }
+                        t = m_lexer.Next();
+                    }
+                }
+                expr.operand1 = std::make_shared<ast::ModifiablePrimary>(mp);
+                return std::make_shared<ast::Expression>(expr);
+            }
+        }
+        ast::Expression expr;
+        ast::Primitive prim;
+        prim.value = tok;
+        expr.operand1 = std::make_shared<ast::Primitive>(prim);
+        return std::make_shared<ast::Expression>(expr);
+    }
     return nullptr;
 }
 
+sPtr<ast::Expression> Parser::parseBinaryExpression(int prec1) {
```

Comment:
Let's give the parameter a sensible default value
```suggestion
sPtr<ast::Expression> Parser::parseBinaryExpression(int prec1 = 0) {
```